File: pa3/backup/List.java
// Chang Kim
// cnkim
// List.java
// pa1

public class List{
   private int index, length;
   private Node front, back, cursor;
   private class Node{
      private Object data;
      private Node next, previous;
      private Node(Object value){
         this.data = value;
         this.next = null;
         this.previous = null;
      }
      public String toString(){
         return this.data.toString();
      }
   }
   public List(){
      this.clear();
   }
   public int length(){
      return this.length;
   }
   public int index(){
      if(this.cursor != null)
         return this.index;
      return -1;
   }
   public Object front(){
      if(this.length > 0)
         return this.front.data;
      return -1;
   }
   public Object back(){
      if(this.length > 0)
         return this.back.data;
      return -1;
   }
   public Object get(){
      if(this.length > 0 && this.index >= 0)
         return this.cursor.data;
      return null;
   }
   public boolean equals(Object L){
      String str1 = this.toString();
      String str2 = L.toString();
      if(str1.equals(str2))
         return true;
      return false;
   }
   public void clear(){
      this.index = -1;
      this.length = 0;
      this.cursor = null;
      this.back = null;
      this.front = null;  
   }
   public void moveFront(){
      if(this.length > 0){
         this.cursor = this.front;
         this.index = 0;
      }
   }
   public void moveBack(){
      if(this.length > 0){
         this.cursor = this.back;
         this.index = this.length-1;
      }
   }
   public void movePrev(){
      if(this.cursor != null){
         if(this.cursor ==  this.front){
            this.cursor = null;
            this.index = -1;
         }
         else{
            this.cursor = this.cursor.previous;
            this.index--;   
         }
      }
   }
   public void moveNext(){
      if(this.cursor != null){
         if(this.cursor == this.back){
            this.cursor = null;
            this.index = -1;
         }
         else{
            this.cursor = this.cursor.next;
            this.index++;
         }
      }
   }
   public void prepend(Object data){
      Node newNode = new Node(data);
      if(this.length == 0){
         this.front = newNode;
         this.back = newNode;
      }
      else{
         newNode.next = this.front;
         this.front.previous = newNode;
         this.front = newNode;
      }
      this.length++;
      if(this.index != -1)
         this.index++;
   }
   public void append(Object data){
      Node newNode = new Node(data);
      if(this.length == 0){
         this.front = newNode;
         this.back = newNode;
      }
      else{
         newNode.previous = this.back;
         this.back.next = newNode;
         this.back = newNode;
      }
      this.length++;
   }
   public void insertBefore(Object data){
      if(this.length > 0){
         if(this.index == 0)
            this.prepend(data);
         else{
            Node newNode = new Node(data);
            newNode.next = this.cursor;
            newNode.previous = this.cursor.previous;
            this.cursor.previous.next = newNode;
            this.cursor.previous = newNode;
            this.length++;
            this.index++;
         }
      }
      else{
         this.append(data);
      }
   }
   public void insertAfter(Object data){
      if(this.length > 0){
         if(this.index == this.length-1)
            this.append(data);
         else{
            Node newNode = new Node(data);
            newNode.previous = this.cursor;
            newNode.next = this.cursor.next;
            this.cursor.next.previous = newNode;
            this.cursor.next = newNode;
            this.length++;
         }
      }
      else{
         this.append(data);
      }
   }
   public void deleteFront(){
      if(this.length > 0){
         if(this.cursor == this.front){
            this.cursor = null;
            this.index = -1;
         }
         else if (this.cursor != null){
            this.index--;
         }
         this.front.next.previous = null;
         this.front = this.front.next;
         this.length--;
      }
   }
   public void deleteBack(){
      if(this.length > 0){
         if(this.length == 1){
            this.back = null;
            this.front = null;
         }
         else{
            this.back.previous.next = null;
            this.back = this.back.previous;
         }
         if(this.index == this.length-1){
            this.cursor = null;
            this.index = -1;
         }
         this.length--;
      }
   }
   public void delete(){
      if(this.length > 0){
         if(this.cursor == this.back)
            this.deleteBack();
         else if(this.cursor == this.front)
            this.deleteFront();
         else{
            this.cursor.previous.next = this.cursor.next;
            this.cursor.next.previous = this.cursor.previous;
            this.cursor = null;
            this.length--;
            this.index = -1;
         }
      }
   }
   public String toString(){
      Node current = this.front;
      String str = "";
      while(current != null){
         str += current.toString() + " ";
         current = current.next;
      }
      return str;
   }
}

File: pa4/List.c
/*
 * List.c
 * Chang Kim
 * cnkim
 * pa2
 */
// Doubly linked list ADT 


#include<stdio.h>
#include<stdlib.h>
#include "List.h"

typedef struct NodeObj{
   int data;
   struct NodeObj* next;
   struct NodeObj* previous;
} NodeObj;

typedef NodeObj* Node;

typedef struct ListObj{
   Node cursor;
   Node front;
   Node back;
   int length;
   int index;
} ListObj;

Node newNode(int data){
   Node N = malloc(sizeof(NodeObj));
   N->data = data;
   N->next = NULL;
   N->previous = NULL;
   return (N);
}

void freeNode(Node* pN){
   if(pN!=NULL && *pN!=NULL){
      free(*pN);
      *pN = NULL;
   }
}

List newList(void){
   List L;
   L = malloc(sizeof(ListObj));
   L->cursor = NULL;
   L->front = NULL;
   L->back = NULL;
   L->length = 0;
   L->index = -1;
   return L;
}
void freeList(List* pL){
   if(pL!=NULL && *pL!=NULL){
      clear(*pL);
      free(*pL);
      *pL = NULL;
   }
}

int length(List L){
   if(L==NULL){
      printf("List Error: NULL list\n");
      exit(1);
   }
   return(L->length);
}

int index(List L){
   if(L==NULL){
      printf("List Error: NULL List reference\n");
      exit(1);
   }
   return (L->index);
}

int front(List L){
   if(L==NULL){
      printf("List Error: calling front() on NULL List reference\n");
      exit(1);
   }
   if(L->length == 0){
      printf("List Error: calling empty List\n");
      exit(1);
   }
   return (L->front->data);
}

int back(List L){
   if(L==NULL){
      printf("List Error: NULL List reference\n");
      exit(1);
   }
   if(L->length == 0){
      printf("List Error: calling empty List\n");
      exit(1);
   }
   return (L->back->data);
}

int get(List L){
   if(L==NULL){
      printf("List Error: NULL List Reference\n");
      exit(1);
   }
   if(L->cursor==NULL){
      printf("List Error: NULL cursor\n");
      exit(1);
   }
   return (L->cursor->data);
}

int equals(List A, List B){
   if(A==NULL || B==NULL){
      printf("List Error: NULL List reference\n");
      exit(1);
   }
   int eq = 0;
   Node N = NULL;
   Node M = NULL;
   eq = (A->length == B->length);
   N = A->front;
   M = B->front;
   while(eq && N!=NULL){
      eq = (N->data == M->data);
      N = N->next;
      M = M->next;
   }
   return eq;
}

void clear(List L){
   if(L==NULL){
      printf("ListError: NULL List reference\n");
      exit(1);
   }
   while(L->length != 0){
      deleteFront(L);
   }
   L->length = 0;
   L->index = -1;
}

void moveFront(List L){
   if(L==NULL){
      printf("List Error: NULL List reference\n");
      exit(1);
   }
   L->cursor = L->front;
   if(L->length != 0){
      L->index = 0;
   }
}

void moveBack(List L){
   if(L==NULL){
      printf("List Error: NULL List reference\n");
      exit(1);
   }
   if(L->length==0){
      printf("List Error: Empty List\n");
      exit(1);
   } 
   L->cursor = L->back;
   L->index = L->length-1;
}

void movePrev(List L){
   if(L==NULL){
      printf("List Error: NULL List reference\n");
      exit(1);
   }
   if(L->cursor==NULL){
      printf("List Error: NULL cursor\n");
      exit(1);
   }
   if(L->cursor==L->front){
      L->cursor = NULL;
      L->index = -1;
   }
   else{
      L->cursor = L->cursor->previous;
      L->index--;
   }
}

void moveNext(List L){
   if(L==NULL){
      printf("List Error: NULL List reference\n");
      exit(1);
   }
   if(L->cursor==NULL){
      printf("List Error: NULL cursor\n");
      exit(1);
   }
   if(L->cursor==L->back){
      L->cursor = NULL;
      L->index = -1;
   }
   else{
      L->cursor = L->cursor->next;
      L->index++;
   }
}

void prepend(List L, int data){
   if(L==NULL){
      printf("List Error: NULL List reference\n");
      exit(1);
   }
   Node N = newNode(data);
   if(L->length==0){
      L->back = N;
   }
   else{
      N->next = L->front;
      L->front->previous = N;
   }
   L->front = N;
   L->length++;
   if(L->index!=-1){
      L->index++;
   }
}

void append(List L, int data){
   if(L==NULL){
      printf("List Error: NULL List reference\n");
      exit(1);
   }
   Node N = newNode(data);
   if(L->length==0){
      L->front = N;
   }
   else{
      N->previous = L->back;
      L->back->next = N;
   }
   L-> back = N;
   L->length++;
}

void insertBefore(List L, int data){
   if(L==NULL){
      printf("List Error: NULL List reference\n");
      exit(1);
   }
   if(L->length==0){
      append(L,data);
      return;
   }
   if(L->index==0){
      prepend(L,data);
   }
   else{
      Node N = newNode(data);
      N->next = L->cursor;
      N->previous = L->cursor->previous;
      L->cursor->previous->next = N;
      L->cursor->previous = N;
      L->length++;
      L->index++;
   }
}

void insertAfter(List L, int data){
   if(L==NULL){
      printf("List Error: NULL List reference\n");
      exit(1);
   }
   if(L->length==0 || L->index==L->length-1){
      append(L,data);
   }
   else{
      Node N = newNode(data);
      N->previous = L->cursor;
      N->next = L->cursor->next;
      L->cursor->next->previous = N;
      L->cursor->next = N;
      L->length++;
   }
}
void deleteFront(List L){
   if(L==NULL){
      printf("List Error: NULL List reference\n");
      exit(1);
   }
   if(L->length==0){
      printf("List Error: Empty List\n");
      exit(1);
   } 
   Node N = NULL;
   N = L->front;
   if(L->length==1){
      L->back = NULL;
      L->front = NULL;
   }
   else{
      L->front = L->front->next;
      L->front->previous = NULL;
   }
   if(L->index==0){
      L->cursor->previous = NULL;
      L->cursor->next = NULL;
      L->cursor = NULL;
      L->index = -1;
   }
   else if(L->cursor!=NULL){
      L->index--;
   }
   L->length--;
   freeNode(&N);
}

void deleteBack(List L){
   if(L==NULL){
      printf("List Error: NULL List reference\n");
      exit(1);
   }
   if(L->length==0){
      printf("List Error: Empty List\n");
      exit(1);
   }
   Node N = NULL;
   N = L->back;
   if(L->length==1){
      L->front = NULL;
      L->back = NULL;
   }
   else{
      L->back = L->back->previous;
      L->back->next = NULL;
   }
   if(L->index==L->length-1){
      L->cursor->next = NULL;
      L->cursor->previous = NULL;
      L->cursor = NULL;
      L->index = -1;
   }
   L->length--;
   freeNode(&N);
}

void delete(List L){
   if(L==NULL){
      printf("List Error: NULL List reference\n");
      exit(1);
   }
   if(L->length==0){
      printf("List Error: List is empty\n");
      exit(1);
   }
   if(L->cursor==NULL){
      printf("List Error: NULL cursor\n");
      exit(1);
   }
   if(L->index==L->length-1){
      deleteBack(L);
   }
   else if(L->index==0){
      deleteFront(L);
   }
   else{
      Node N = NULL;
      N = L->cursor;
      L->cursor->previous->next = L->cursor->next;
      L->cursor->next->previous = L->cursor->previous;
      L->length--;
      L->index = -1;
      freeNode(&N); 
   }
}

void printList(FILE* out, List L){
   if(L==NULL){
      printf("List Error: NULL List reference");
      exit(1);
   }
   Node N = NULL;
   for(N=L->front; N!=NULL; N=N->next){
      fprintf(out,"%d ", N->data);
   }
}

List copyList(List L){
   if(L==NULL){
      printf("List Error: NULL List reference");
      exit(1);
   }
   List L2 = newList();
   Node N = NULL;
   for(N=L->front; N!=NULL; N=N->next){
      append(L2,N->data);
   }
   return L2;
}
